### PR TITLE
Enhance the algorithm of merkle tree to fix errors in boundary conditions.

### PR DESCRIPTION
**Pull request by chaoticlonghair**
* Enhance the algorithm of merkle tree to fix errors in boundary conditions.

    When execute [`MerkleTree::get_proof_by_input_index`](https://github.com/cryptape/cita-common/blob/6a9be3c62a50abda9df7571b5ec8ef53e7cd219e/util/src/merklehash.rs#L57-L67), if

  1. `input_index` is over `input_size`

  or

  2. `input_size` is `zero`

  , there will be some unexpected errors.

* Add some boundary tests.

* Rename several variables / functions to increase readability.

References:
* [PR-34: Rewrite the algorithm of merkle tree](https://github.com/cryptape/cita-common/pull/34)

### CODE REVIEW

**Comment by volzkzg** [comment]
get_size_of_nodes 这边的 m 应该已经不会为 0 了吧。

**Review by volzkzg — APPROVED** [review]

**Comment by volzkzg** [comment]
有个问题，一定要最后一层存 A-F，倒数第二层存 G.. 吗？如果直接按照顺序 A.. 存下来会好写并且简洁很多。比如像[这样子](https://gist.github.com/rust-play/cb80395840c00a59abc6656594bfeb4e)。

**Comment by chaoticlonghair** [comment]
@volzkzg 
For being compatible with the old version.
[Here is the old tests](https://github.com/cryptape/cita-common/blob/de92b66c2d7e9b7079f63934f5fdbf9fc560b782/util/src/merklehash.rs#L93-L153).

File: util/src/merklehash.rs [hunk 1]
```diff
@@ -154,12 +157,14 @@ fn generate_merkle_tree_from_hash(input: Vec<H256>) -> MerkleTree {
...
 // Calc hash for one row in merkle tree.
 // If break_cnt > 0, just calc break_cnt hash for that row.
-fn calc_merkle_tree_at_row(nodes: &mut Vec<H256>, row: usize, break_cnt: usize) {
-    let index_update = (1 << (row - 1)) - 1;
-    let size_update = if break_cnt > 0 {
+fn calc_merkle_tree_at_row(nodes: &mut Vec<H256>, row_index: usize, break_cnt: usize) {
+    // The first index in the row which above the row_index row.
+    let index_update = (1 << (row_index - 1)) - 1;
+    let size_max = 1 << (row_index - 1);
+    let size_update = if break_cnt > 0 && break_cnt < size_max {
```
Comment (chaoticlonghair) [inline util/src/merklehash.rs:164]: For safety, check if `break_cnt` is overflow.

File: util/src/merklehash.rs [hunk 2]
```diff
@@ -113,10 +115,11 @@ fn generate_merkle_tree_from_hash(input: Vec<H256>) -> MerkleTree {
             let nodes_size = get_size_of_nodes(input_size);
             let mut nodes = vec![H256::default(); nodes_size];
 
-            let lowest_row = get_lowest_row(input_size);
+            let rows_size = get_size_of_rows(input_size);
```
Comment (chaoticlonghair) [inline util/src/merklehash.rs:118]: Rename to increase readability.

File: util/src/merklehash.rs [hunk 3]
```diff
@@ -187,61 +192,77 @@ fn get_lowest_row(m: usize) -> usize {
...
     //      n1 = 2 ^ y - 1;
-    // Size of nodes for the lowest row:
+    // Size of nodes in the lowest row:
     //      n2 = (m - z) * 2;
     // Returns:
     //      n1 + n2
     //      = (2 ^ y - 1) + ((m - z) * 2)
     //      = m * 2  - 1
-    m * 2 - 1
+    if m == 0 {
```
Comment (chaoticlonghair) [inline util/src/merklehash.rs:209]: For safety, check if `m` is overflow.

File: util/src/merklehash.rs [hunk 4]
```diff
@@ -187,61 +192,77 @@ fn get_lowest_row(m: usize) -> usize {
...
+    } else {
+        m * 2 - 1
+    }
 }
 
-// Use math index (start from 1).
 #[inline]
-fn get_math_index_of_brother_node(mi: usize) -> usize {
-    (mi & ((!0) - 1)) + ((!mi) & 1)
+fn get_index_of_brother_and_father(index: usize) -> (usize, usize) {
```
Comment (chaoticlonghair) [inline util/src/merklehash.rs:217]: Rename and add some comments to increase readability.

**Review by u2 — APPROVED** [review]

**Pull request merged**